src/Compilers/CSharp/Portable/Symbols/Source/LocalFunctionSymbol.cs

```diff
@@ -289,16 +289,16 @@ public override bool IsExtensionMethod
         // Replace with IsStatic after fixing https://github.com/dotnet/roslyn/issues/27719.
         internal bool IsStaticLocalFunction => _syntax.Modifiers.Any(SyntaxKind.StaticKeyword);
 
-        internal override TypeSymbol IteratorElementType
+        internal override TypeSymbolWithAnnotations IteratorElementType
         {
             get
             {
                 return _iteratorElementType;
             }
             set
             {
-                Debug.Assert((object)_iteratorElementType == null || TypeSymbol.Equals(_iteratorElementType, value, TypeCompareKind.ConsiderEverything2));
-                Interlocked.CompareExchange(ref _iteratorElementType, value, null);
+                Debug.Assert(_iteratorElementType.IsDefault || TypeSymbol.Equals(_iteratorElementType.TypeSymbol, value.TypeSymbol, TypeCompareKind.ConsiderEverything2));
+                _iteratorElementType = value;
```

Comment:
>_iteratorElementType = value; [](start = 16, length = 29)

The assignment is not thread-safe.

`_iteratorElementType` should probably be a `TypeSymbolWithAnnotations.Builder`. 

Then this property would be:
```
get => _iteratorElementType.ToType();
set => _iteratorElementType.InterlockedInitialize(value);
```